### PR TITLE
Bugfix:  don't match account numbers as cfn ids

### PR DESCRIPTION
**Pull request by andrewjhumphrey**
Liam reported that buckets like  elasticbeanstalk-us-east-1-298865909318
we incorrectly excluded by iamy on the basis that they matched the cfn heuristic regexp

This change is a little more involved than it first seems it should be as it required installing a new 
regular expression library that supports lookaheads.   Golang's native regexp package
doesn't support then as they aren't O(n) complexity.

If anyone can rewrite the lookahead regexp to not require lookahead, please feel free.

### CODE REVIEW

File: iamy/cfn.go [hunk 1]
```diff
@@ -13,7 +13,7 @@ import (
 	"github.com/aws/aws-sdk-go/service/cloudformation/cloudformationiface"
 )
 
-var cfnResourceRegexp = regexp.MustCompile(`-[A-Z0-9]{10,20}$`)
+var cfnResourceRegexp = regexp.MustCompile(`^(?=.{10,20})(?=.*[A-Z])(?=.*[0-9A-Z]).*$`)
```
Comment (stevehodgkiss) [inline iamy/cfn.go]: Shall we add an additional test? https://github.com/envato/iamy/blob/e13a87f129bef1aa17fc2c776e09c49b32713f55/iamy/cfn_test.go#L30-L32

Comment (andrewjhumphrey) [reply iamy/cfn.go]: Yep!  Doing that now.

Comment (andrewjhumphrey) [reply iamy/cfn.go]: 521f98d

File: go.sum [hunk 2]
```diff
@@ -14,6 +14,8 @@ github.com/fatih/color v1.7.0 h1:DkWD4oS2D8LGGgTQ6IvwJJXSL5Vp2ffcQg58nFV38Ys=
 github.com/fatih/color v1.7.0/go.mod h1:Zm6kSWBoL9eyXnKyktHP6abPY2pDugNf5KwzbycvMj4=
 github.com/ghodss/yaml v1.0.0 h1:wQHKEahhL6wmXdzwWG11gIVCkOv05bNOh+Rxn0yngAk=
 github.com/ghodss/yaml v1.0.0/go.mod h1:4dBDuWmgqj2HViK6kFavaiC9ZROes6MMH2rRYeMEF04=
+github.com/h2so5/goback v0.0.0-20150302055225-6e210305bfc9 h1:HpHh/z38D53+rJJnOZhVGcaPImRngX18xKVma4VHBEs=
+github.com/h2so5/goback v0.0.0-20150302055225-6e210305bfc9/go.mod h1:YaAwP3MyryOTskva8K5dvyse+hrwc/OwgP4rHXcGS98=
```
Comment (stevehodgkiss) [inline go.sum]: Is this change needed? I get a 404 at https://github.com/h2so5/goback

File: iamy/cfn.go (hunk 1 above)
Comment (stevehodgkiss) [inline iamy/cfn.go]: I think it'd be better to have a separate regex for elasticbeanstalk managed buckets, and have a separate check in https://github.com/envato/iamy/blob/main/iamy/cfn.go#L138-L140

It looks like they're in the format of `elasticbeanstalk-$region-$accountid`. Also wondering if the matcher for this should be tightened up to match the initial `elasticbeanstalk-` as well as the `-$accountid`... https://docs.aws.amazon.com/elasticbeanstalk/latest/dg/AWSHowTo.S3.html

File: go.sum (hunk 2 above)
Comment (andrewjhumphrey) [reply go.sum]: It's the results of me doing a 
go get github.com/h2so5/goback/regexp

Does that work for you?

File: iamy/cfn.go (hunk 1 above)
Comment (andrewjhumphrey) [reply iamy/cfn.go]: Easy enough, not entirely sure it's worth trying to support every weird s3 bucket naming scheme we come across.

Things ending in account numbers are pretty common (we've used the scheme in several places).
I'm not trying to change this to detect beanstalk buckets, I'm trying to make sure we don't capture account numbers when we go looking for a cfn id suffix...

Comment (andrewjhumphrey) [reply iamy/cfn.go]: So even if we specifically exclude the beanstalk bucket in this example we'll still need to fix the regular expression as it catches things that are not cfn suffix ids.   Beanstalk is just the one we happened to notice..

File: go.sum (hunk 2 above)
Comment (stevehodgkiss) [reply go.sum]: `go get` on this branch works. What are we using from that library in this PR?

File: iamy/cfn.go (hunk 1 above)
Comment (stevehodgkiss) [reply iamy/cfn.go]: ~I imagine a separate regex for `endsInAccountId` could be created, then a check earlier than the CFN regex would `return true` early. That should mean the cfn regex wouldn't need updating. The main goal of that is to have simpler regex's so they're easier to understand.~

Comment (andrewjhumphrey) [reply iamy/cfn.go]: So far I've found
beanstalk, code-star, config, athena, sagemaker all using the same scheme 
as well as the envato-logs-<accountnumber> stuff we devised

File: iamy/cfn.go [hunk 3]
```diff
@@ -135,7 +136,7 @@ func (c *cfnClient) IsManagedResource(cfnType CfnResourceType, resourceIdentifie
 		return c.managedResources[resourceIdentifier].contains(cfnType)
 	}
 
-	if cfnResourceRegexp.MatchString(resourceIdentifier) {
+	if cfnResourceRegexp.MatchString(resourceIdentifier) && strings.ContainsAny(resourceIdentifier, UpperCaseLetters) {
```
Comment (stevehodgkiss) [inline iamy/cfn.go]: We could limit the `ContainsAny` check to the last component of the resourceIdentifier (after the last `-`). It probably doesn't make a difference for the pattern we're looking to exclude though.

**Review by stevehodgkiss — DISMISSED** [review]

File: iamy/cfn.go (hunk 3 above)
Comment (andrewjhumphrey) [reply iamy/cfn.go]: I think it will sadly, I've pushed a fix and a couple of tests to catch it.

Comment (andrewjhumphrey) [reply iamy/cfn.go]: https://github.com/envato/iamy/pull/11/commits/5635ecd4361b146e4d365ca9c67243e795db3107

File: iamy/cfn_test.go [hunk 4]
```diff
@@ -30,5 +30,17 @@ func TestCfnMangedResources(t *testing.T) {
 		if !cfn.IsManagedResource(CfnIamPolicy, "foobar-ABCDEFGH1234567") {
 			t.Fatal("names with id suffix are managed")
 		}
+
+		if cfn.IsManagedResource(CfnIamPolicy, "foobar-abcdefgh1234567") {
+			t.Fatal("names with suffix containing only lowercase letters are not managed")
```
Comment (patrobinson) [inline iamy/cfn_test.go:35]: I find the language of this tests to be quite confusing, but realise that's not a problem created here :)

**Review by patrobinson — APPROVED** [review]

File: go.sum (hunk 2 above)
Comment (andrewjhumphrey) [reply go.sum]: Nothing anymore :smile:  I removed it again when we switched away from the complex regular expression.

File: iamy/cfn_test.go (hunk 4 above)
Comment (andrewjhumphrey) [reply iamy/cfn_test.go:35]: Work for future PRs.
